lib/DxilPIXPasses/DxilDbgValueToDbgDeclare.cpp

```diff
@@ -0,0 +1,740 @@
+    //
+    //     CurrentAlignedOffset = 0x00000048(72)
+    //
+    // Then
+    //
+    //     T = CurrentAlignOffset + AlignMask = 0x00000057(87)
+    //
+    // Which mean
+    //
+    //     T & ~CurrentOffset = 0x00000050(80)
+    //
+    // is the aligned offset where Ty should be placed.
+    const unsigned AlignMask = Ty->getAlignInBits() - 1;
+    m_CurrentAlignedOffset =
+        (m_CurrentAlignedOffset + AlignMask) & ~AlignMask;
+  }
+
+  // Add is used to "add" an aggregate element (struct field, array element)
+  // at the current aligned/packed offsets, bumping them by Ty's size.
+  OffsetInBits Add(
+      llvm::DIBasicType *Ty
+  )
+  {
+    m_PackedOffsetToAlignedOffset[m_CurrentPackedOffset] = m_CurrentAlignedOffset;
+    m_AlignedOffsetToPackedOffset[m_CurrentAlignedOffset] = m_CurrentPackedOffset;
+
+    const OffsetInBits Ret = m_CurrentAlignedOffset;
+    m_CurrentPackedOffset += Ty->getSizeInBits();
+    m_CurrentAlignedOffset += Ty->getSizeInBits();
+
+    return Ret;
+  }
+
+  // AlignToAndAddUnhandledType is used for error handling when Ty
+  // could not be handled by the transformation. This is a best-effort
+  // way to continue the pass by ignoring the current type and hoping
+  // that adding Ty as a blob other fields/elements added will land
+  // in the proper offset.
+  void AlignToAndAddUnhandledType(
+      llvm::DIType *Ty
+  )
+  {
+    AlignTo(Ty);
+    m_CurrentPackedOffset += Ty->getSizeInBits();
+    m_CurrentAlignedOffset += Ty->getSizeInBits();
+  }
+
+  bool GetAlignedOffsetFromPackedOffset(
+      OffsetInBits PackedOffset,
+      OffsetInBits *AlignedOffset
+  ) const
+  {
+    return GetOffsetWithMap(
+        m_PackedOffsetToAlignedOffset,
+        PackedOffset,
+        AlignedOffset);
+  }
+
+  OffsetInBits GetPackedOffsetFromAlignedOffset(
+      OffsetInBits AlignedOffset,
+      OffsetInBits *PackedOffset
+  ) const
+  {
+    return GetOffsetWithMap(
+        m_PackedOffsetToAlignedOffset,
+        AlignedOffset,
+        PackedOffset);
+  }
+
+  OffsetInBits GetCurrentPackedOffset() const
+  {
+    return m_CurrentPackedOffset;
+  }
+
+  OffsetInBits GetCurrentAlignedOffset() const
+  {
+      return m_CurrentAlignedOffset;
+  }
+
+private:
+  OffsetInBits m_CurrentPackedOffset = 0;
+  OffsetInBits m_CurrentAlignedOffset = 0;
+
+  using OffsetMap = std::unordered_map<OffsetInBits, OffsetInBits>;
+
+  OffsetMap m_PackedOffsetToAlignedOffset;
+  OffsetMap m_AlignedOffsetToPackedOffset;
+
+  static bool GetOffsetWithMap(
+      const OffsetMap &Map,
+      OffsetInBits SrcOffset,
+      OffsetInBits *DstOffset
+  )
+  {
+    auto it = Map.find(SrcOffset);
+    if (it == Map.end())
+    {
+      return false;
+    }
+
+    *DstOffset = it->second;
+    return true;
+  }
+};
+
+// VariableRegisters contains the logic for traversing a DIType T and
+// creating AllocaInsts that map back to a specific offset within T.
+class VariableRegisters
+{
+public:
+  explicit VariableRegisters(
```

Comment:
(I don;t think explicit does anything if there are more than one argument to the constructor :=) )

Reply:
I really wish the compiler would at least warn about this... :)